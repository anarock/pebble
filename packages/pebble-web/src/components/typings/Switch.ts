import { Interpolation } from "./emotionCustom";

export interface SwitchProps {
  initialValue?: boolean;
  onChange?: (value: boolean) => void;
  styles?: Interpolation;
  label?: string;
  disabled?: boolean;
}

export interface SwitchState {
  value: boolean;
}
