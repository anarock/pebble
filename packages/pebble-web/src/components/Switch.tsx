import * as React from "react";

import {
  labelTextStyle,
  labelStyle,
  selectedLabel,
  inputStyle,
  selectedSwitch,
  switchStyle,
  fixedLabelStyle,
  disabledStyle
} from "./styles/Switch.styles";
import { SwitchProps, SwitchState } from "./typings/Switch";

export default class Switch extends React.PureComponent<
  SwitchProps,
  SwitchState
> {
  state = {
    value: this.props.initialValue || false
  };

  render() {
    const { onChange, styles, label, disabled } = this.props;
    const { value } = this.state;

    return (
      <label css={[styles, fixedLabelStyle, !!disabled && disabledStyle]}>
        <span css={labelTextStyle}>{label}</span>
        <div css={[labelStyle, value && selectedLabel]}>
          <input
            type="checkbox"
            checked={value}
            css={inputStyle}
            onChange={() => {
              this.setState({
                value: !value
              });
              if (onChange) {
                onChange(!value);
              }
            }}
            disabled={disabled}
          />
          <span css={[switchStyle, value && selectedSwitch]} />
        </div>
      </label>
    );
  }
}
