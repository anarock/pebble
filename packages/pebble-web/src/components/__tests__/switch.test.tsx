import * as React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";
import Switch from "../Switch";

describe("Component: Switch", () => {
  test("snapshot", () => {
    const switch1 = shallow(<Switch label="I am a switch" />);
    expect(switch1).toMatchSnapshot();

    const switch2 = shallow(<Switch label="I am a switch" />);
    expect(switch2).toMatchSnapshot();
  });

  test("onChange is triggered on click", () => {
    const spy = sinon.spy();

    const switch1 = mount(<Switch label="I am a switch" onChange={spy} />);
    switch1.find("input").simulate("change");
    expect(spy.calledOnce).toBeTruthy();
  });
});
