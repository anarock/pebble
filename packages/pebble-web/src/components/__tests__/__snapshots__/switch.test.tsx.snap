// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Switch snapshot 1`] = `
.emotion-0 {
  font-weight: 500;
  color: #6B7785;
  font-size: 12px;
  margin-right: 10px;
}

.emotion-3 {
  width: 28px;
  height: 14px;
  background-color: #9FA6B0;
  position: relative;
  display: inline-block;
  border-radius: 14px;
  cursor: pointer;
}

.emotion-4 {
  padding: 15px 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-1 {
  width: 0;
  height: 0;
  opacity: 0;
}

.emotion-2 {
  position: absolute;
  height: 10px;
  width: 10px;
  left: 2px;
  bottom: 2px;
  border-radius: 50%;
  background-color: #ffffff;
  -webkit-transition: .4s;
  transition: .4s;
}

<label
  className="emotion-4"
>
  <span
    className="emotion-0"
  >
    I am a switch
  </span>
  <div
    className="emotion-3"
  >
    <input
      checked={false}
      className="emotion-1"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="emotion-2"
    />
  </div>
</label>
`;

exports[`Component: Switch snapshot 2`] = `
.emotion-0 {
  font-weight: 500;
  color: #6B7785;
  font-size: 12px;
  margin-right: 10px;
}

.emotion-3 {
  width: 28px;
  height: 14px;
  background-color: #9FA6B0;
  position: relative;
  display: inline-block;
  border-radius: 14px;
  cursor: pointer;
}

.emotion-4 {
  padding: 15px 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-1 {
  width: 0;
  height: 0;
  opacity: 0;
}

.emotion-2 {
  position: absolute;
  height: 10px;
  width: 10px;
  left: 2px;
  bottom: 2px;
  border-radius: 50%;
  background-color: #ffffff;
  -webkit-transition: .4s;
  transition: .4s;
}

<label
  className="emotion-4"
>
  <span
    className="emotion-0"
  >
    I am a switch
  </span>
  <div
    className="emotion-3"
  >
    <input
      checked={false}
      className="emotion-1"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="emotion-2"
    />
  </div>
</label>
`;
